File: api_client.js
/**
 * AWS Trivia Questions API Client - JavaScript/Node.js SDK
 * 
 * Usage:
 *   // Browser
 *   const client = new AWSTriviaAPIClient('https://your-api-domain.com');
 *   
 *   // Node.js
 *   const AWSTriviaAPIClient = require('./api_client.js');
 *   const client = new AWSTriviaAPIClient('https://your-api-domain.com');
 */

class AWSTriviaAPIClient {
    /**
     * Initialize the API client
     * @param {string} baseUrl - Base URL of the API server
     * @param {string} apiVersion - API version to use (default: v1)
     */
    constructor(baseUrl, apiVersion = 'v1') {
        this.baseUrl = baseUrl.replace(/\/$/, '');
        this.apiVersion = apiVersion;
        this.apiBase = `${this.baseUrl}/api/${apiVersion}`;
        
        // Detect environment (Node.js vs Browser)
        this.isNode = typeof window === 'undefined';
        
        if (this.isNode) {
            // Node.js environment
            try {
                this.fetch = require('node-fetch');
            } catch (e) {
                console.warn('node-fetch not found. Please install: npm install node-fetch');
                this.fetch = null;
            }
        } else {
            // Browser environment
            this.fetch = window.fetch.bind(window);
        }
    }
    
    /**
     * Make HTTP request to API endpoint
     * @param {string} endpoint - API endpoint
     * @param {Object} params - Query parameters
     * @returns {Promise<Object>} API response
     */
    async _makeRequest(endpoint, params = {}) {
        try {
            if (!this.fetch) {
                throw new Error('Fetch not available. In Node.js, install node-fetch: npm install node-fetch');
            }
            
            const url = new URL(`${this.apiBase}${endpoint}`);
            
            // Add query parameters
            Object.keys(params).forEach(key => {
                if (params[key] !== null && params[key] !== undefined) {
                    url.searchParams.append(key, params[key]);
                }
            });
            
            const response = await this.fetch(url.toString(), {
                method: 'GET',
                headers: {
                    'Content-Type': 'application/json',
                    'User-Agent': 'AWS-Trivia-API-Client-JS/1.0'
                }
            });
            
            if (!response.ok) {
                throw new Error(`HTTP ${response.status}: ${response.statusText}`);
            }
            
            const data = await response.json();
            return data;
            
        } catch (error) {
            return {
                success: false,
                error: 'Request failed',
                message: error.message,
                timestamp: new Date().toISOString()
            };
        }
    }
    
    /**
     * Check API health status
     * @returns {Promise<Object>} Health status
     */
    async healthCheck() {
        return await this._makeRequest('/health');
    }
    
    /**
     * Get API information and statistics
     * @returns {Promise<Object>} API information
     */
    async getApiInfo() {
        return await this._makeRequest('/info');
    }
    
    /**
     * Get questions with optional filtering and pagination
     * @param {Object} options - Query options
     * @param {number} options.limit - Number of questions to return (max 1000)
     * @param {number} options.offset - Pagination offset
     * @param {number} options.level - Filter by level (1-5)
     * @param {string} options.category - Filter by category
     * @param {string} options.difficulty - Filter by difficulty
     * @param {boolean} options.randomize - Randomize results
     * @returns {Promise<Object>} Questions response
     */
    async getQuestions(options = {}) {
        const {
            limit = 10,
            offset = 0,
            level,
            category,
            difficulty,
            randomize = false
        } = options;
        
        const params = {
            limit,
            offset,
            random: randomize.toString()
        };
        
        if (level !== undefined) params.level = level;
        if (category) params.category = category;
        if (difficulty) params.difficulty = difficulty;
        
        return await this._makeRequest('/questions', params);
    }
    
    /**
     * Get random questions with optional filtering
     * @param {Object} options - Query options
     * @param {number} options.count - Number of questions to return
     * @param {number} options.level - Filter by level
     * @param {string} options.category - Filter by category
     * @param {string} options.difficulty - Filter by difficulty
     * @returns {Promise<Object>} Random questions response
     */
    async getRandomQuestions(options = {}) {
        const {
            count = 10,
            level,
            category,
            difficulty
        } = options;
        
        const params = { count };
        
        if (level !== undefined) params.level = level;
        if (category) params.category = category;
        if (difficulty) params.difficulty = difficulty;
        
        return await this._makeRequest('/questions/random', params);
    }
    
    /**
     * Get a specific question by ID
     * @param {number} questionId - Question ID
     * @returns {Promise<Object>} Question response
     */
    async getQuestionById(questionId) {
        return await this._makeRequest(`/questions/${questionId}`);
    }
    
    /**
     * Get questions from a specific level
     * @param {number} level - Level number (1-5)
     * @param {Object} options - Query options
     * @param {number} options.limit - Number of questions to return
     * @param {number} options.offset - Pagination offset
     * @param {boolean} options.randomize - Randomize results
     * @returns {Promise<Object>} Level questions response
     */
    async getQuestionsByLevel(level, options = {}) {
        const {
            limit = 100,
            offset = 0,
            randomize = false
        } = options;
        
        const params = {
            limit,
            offset,
            random: randomize.toString()
        };
        
        return await this._makeRequest(`/questions/level/${level}`, params);
    }
    
    /**
     * Get questions from a specific category
     * @param {string} category - Category name
     * @param {Object} options - Query options
     * @param {number} options.limit - Number of questions to return
     * @param {number} options.offset - Pagination offset
     * @param {boolean} options.randomize - Randomize results
     * @returns {Promise<Object>} Category questions response
     */
    async getQuestionsByCategory(category, options = {}) {
        const {
            limit = 100,
            offset = 0,
            randomize = false
        } = options;
        
        const params = {
            limit,
            offset,
            random: randomize.toString()
        };
        
        return await this._makeRequest(`/questions/category/${category}`, params);
    }
    
    /**
     * Get questions from a specific difficulty level
     * @param {string} difficulty - Difficulty level
     * @param {Object} options - Query options
     * @param {number} options.limit - Number of questions to return
     * @param {number} options.offset - Pagination offset
     * @param {boolean} options.randomize - Randomize results
     * @returns {Promise<Object>} Difficulty questions response
     */
    async getQuestionsByDifficulty(difficulty, options = {}) {
        const {
            limit = 100,
            offset = 0,
            randomize = false
        } = options;
        
        const params = {
            limit,
            offset,
            random: randomize.toString()
        };
        
        return await this._makeRequest(`/questions/difficulty/${difficulty}`, params);
    }
    
    /**
     * Get all available categories
     * @returns {Promise<Object>} Categories response
     */
    async getCategories() {
        return await this._makeRequest('/categories');
    }
    
    /**
     * Get all available difficulty levels
     * @returns {Promise<Object>} Difficulties response
     */
    async getDifficulties() {
        return await this._makeRequest('/difficulties');
    }
    
    /**
     * Get all available game levels
     * @returns {Promise<Object>} Levels response
     */
    async getLevels() {
        return await this._makeRequest('/levels');
    }
    
    /**
     * Export all questions as JSON
     * @param {boolean} includeAnswers - Whether to include correct answers
     * @returns {Promise<Object>} Export response
     */
    async exportAllQuestions(includeAnswers = true) {
        const params = {
            include_answers: includeAnswers.toString()
        };
        
        return await this._makeRequest('/export/json', params);
    }
    
    /**
     * Parse questions from API response
     * @param {Object} response - API response containing questions
     * @returns {Array} Array of question objects
     */
    parseQuestions(response) {
        if (!response.success || !response.data || !response.data.questions) {
            return [];
        }
        
        return response.data.questions.map(q => ({
            id: q.id || 0,
            question: q.question || '',
            options: q.options || [],
            answer: q.answer || 0,
            difficulty: q.difficulty || '',
            category: q.category || '',
            level: q.level || 1,
            levelName: q.level_name || ''
        }));
    }
    
    /**
     * Create a quiz from random questions
     * @param {Object} options - Quiz options
     * @param {number} options.count - Number of questions
     * @param {number} options.level - Filter by level
     * @param {string} options.category - Filter by category
     * @param {string} options.difficulty - Filter by difficulty
     * @returns {Promise<Object>} Quiz object
     */
    async createQuiz(options = {}) {
        const response = await this.getRandomQuestions(options);
        
        if (!response.success) {
            return {
                success: false,
                error: response.error,
                message: response.message
            };
        }
        
        const questions = this.parseQuestions(response);
        
        return {
            success: true,
            quiz: {
                id: Date.now().toString(),
                questions: questions,
                totalQuestions: questions.length,
                createdAt: new Date().toISOString(),
                options: options
            }
        };
    }
}

// Example usage function
async function exampleUsage() {
    console.log('🚀 AWS Trivia API Client - JavaScript Example');
    console.log('='.repeat(50));
    
    const client = new AWSTriviaAPIClient('http://localhost:5001');
    
    try {
        // Health check
        console.log('\n1. Health Check:');
        const health = await client.healthCheck();
        if (health.success) {
            console.log(`✅ API is healthy - ${health.total_questions || 0} questions available`);
        } else {
            console.log(`❌ API health check failed: ${health.error}`);
        }
        
        // Get API info
        console.log('\n2. API Information:');
        const info = await client.getApiInfo();
        if (info.success) {
            console.log(`📊 Total Questions: ${info.data.total_questions || 0}`);
            console.log(`🏷️  Categories: ${info.data.categories?.length || 0}`);
            console.log(`📈 Difficulties: ${info.data.difficulties?.length || 0}`);
            console.log(`🎯 Levels: ${Object.keys(info.data.levels || {}).length}`);
        }
        
        // Get random questions
        console.log('\n3. Random Questions (3):');
        const randomResponse = await client.getRandomQuestions({ count: 3 });
        if (randomResponse.success) {
            const questions = client.parseQuestions(randomResponse);
            questions.forEach((q, i) => {
                console.log(`\n   Question ${i + 1}: ${q.question}`);
                console.log(`   Category: ${q.category} | Difficulty: ${q.difficulty} | Level: ${q.level}`);
                q.options.forEach((option, j) => {
                    const marker = j === q.answer ? '✓' : ' ';
                    console.log(`   ${marker} ${String.fromCharCode(65 + j)}. ${option}`);
                });
            });
        }
        
        // Create a quiz
        console.log('\n4. Create Quiz (Level 1, 2 questions):');
        const quiz = await client.createQuiz({ count: 2, level: 1 });
        if (quiz.success) {
            console.log(`   ✅ Quiz created with ${quiz.quiz.totalQuestions} questions`);
            quiz.quiz.questions.forEach((q, i) => {
                console.log(`   • Question ${i + 1}: ${q.question} (${q.category})`);
            });
        }
        
        // Get categories
        console.log('\n5. Available Categories:');
        const categoriesResponse = await client.getCategories();
        if (categoriesResponse.success) {
            const categories = categoriesResponse.data.categories || [];
            console.log(`   ${categories.join(', ')}`);
        }
        
        console.log('\n' + '='.repeat(50));
        console.log('✅ JavaScript API Client Example Complete!');
        
    } catch (error) {
        console.error('❌ Example failed:', error.message);
    }
}

// Export for different environments
if (typeof module !== 'undefined' && module.exports) {
    // Node.js
    module.exports = AWSTriviaAPIClient;
} else if (typeof window !== 'undefined') {
    // Browser
    window.AWSTriviaAPIClient = AWSTriviaAPIClient;
}

// Run example if this file is executed directly in Node.js
if (typeof require !== 'undefined' && require.main === module) {
    exampleUsage();
}

File: api_client.py
#!/usr/bin/env python3
"""
AWS Trivia Questions API Client
Python SDK for accessing the AWS Trivia Questions API
"""

import requests
import json
from typing import List, Dict, Optional, Union
from dataclasses import dataclass
from datetime import datetime

@dataclass
class Question:
    """Question data class"""
    id: int
    question: str
    options: List[str]
    answer: int
    difficulty: str
    category: str
    level: int
    level_name: str

@dataclass
class APIResponse:
    """API response data class"""
    success: bool
    data: Dict
    timestamp: str
    error: Optional[str] = None
    message: Optional[str] = None

class AWSTrivia APIClient:
    """
    Python client for the AWS Trivia Questions API
    
    Usage:
        client = AWSTriviaAPIClient("https://your-api-domain.com")
        questions = client.get_random_questions(count=10)
    """
    
    def __init__(self, base_url: str, api_version: str = "v1"):
        """
        Initialize the API client
        
        Args:
            base_url: Base URL of the API server
            api_version: API version to use (default: v1)
        """
        self.base_url = base_url.rstrip('/')
        self.api_version = api_version
        self.api_base = f"{self.base_url}/api/{api_version}"
        self.session = requests.Session()
        self.session.headers.update({
            'Content-Type': 'application/json',
            'User-Agent': 'AWS-Trivia-API-Client/1.0'
        })
    
    def _make_request(self, endpoint: str, params: Optional[Dict] = None) -> APIResponse:
        """Make HTTP request to API endpoint"""
        try:
            url = f"{self.api_base}{endpoint}"
            response = self.session.get(url, params=params)
            response.raise_for_status()
            
            data = response.json()
            return APIResponse(
                success=data.get('success', False),
                data=data.get('data', {}),
                timestamp=data.get('timestamp', ''),
                error=data.get('error'),
                message=data.get('message')
            )
            
        except requests.exceptions.RequestException as e:
            return APIResponse(
                success=False,
                data={},
                timestamp=datetime.utcnow().isoformat(),
                error='Request failed',
                message=str(e)
            )
        except json.JSONDecodeError as e:
            return APIResponse(
                success=False,
                data={},
                timestamp=datetime.utcnow().isoformat(),
                error='Invalid JSON response',
                message=str(e)
            )
    
    def health_check(self) -> APIResponse:
        """Check API health status"""
        return self._make_request('/health')
    
    def get_api_info(self) -> APIResponse:
        """Get API information and statistics"""
        return self._make_request('/info')
    
    def get_questions(self, 
                     limit: int = 10, 
                     offset: int = 0,
                     level: Optional[int] = None,
                     category: Optional[str] = None,
                     difficulty: Optional[str] = None,
                     randomize: bool = False) -> APIResponse:
        """
        Get questions with optional filtering and pagination
        
        Args:
            limit: Number of questions to return (max 1000)
            offset: Pagination offset
            level: Filter by level (1-5)
            category: Filter by category
            difficulty: Filter by difficulty
            randomize: Randomize results
        """
        params = {
            'limit': limit,
            'offset': offset,
            'random': str(randomize).lower()
        }
        
        if level is not None:
            params['level'] = level
        if category:
            params['category'] = category
        if difficulty:
            params['difficulty'] = difficulty
        
        return self._make_request('/questions', params)
    
    def get_random_questions(self,
                           count: int = 10,
                           level: Optional[int] = None,
                           category: Optional[str] = None,
                           difficulty: Optional[str] = None) -> APIResponse:
        """
        Get random questions with optional filtering
        
        Args:
            count: Number of questions to return
            level: Filter by level
            category: Filter by category
            difficulty: Filter by difficulty
        """
        params = {'count': count}
        
        if level is not None:
            params['level'] = level
        if category:
            params['category'] = category
        if difficulty:
            params['difficulty'] = difficulty
        
        return self._make_request('/questions/random', params)
    
    def get_question_by_id(self, question_id: int) -> APIResponse:
        """Get a specific question by ID"""
        return self._make_request(f'/questions/{question_id}')
    
    def get_questions_by_level(self, 
                              level: int,
                              limit: int = 100,
                              offset: int = 0,
                              randomize: bool = False) -> APIResponse:
        """
        Get questions from a specific level
        
        Args:
            level: Level number (1-5)
            limit: Number of questions to return
            offset: Pagination offset
            randomize: Randomize results
        """
        params = {
            'limit': limit,
            'offset': offset,
            'random': str(randomize).lower()
        }
        
        return self._make_request(f'/questions/level/{level}', params)
    
    def get_questions_by_category(self,
                                 category: str,
                                 limit: int = 100,
                                 offset: int = 0,
                                 randomize: bool = False) -> APIResponse:
        """
        Get questions from a specific category
        
        Args:
            category: Category name
            limit: Number of questions to return
            offset: Pagination offset
            randomize: Randomize results
        """
        params = {
            'limit': limit,
            'offset': offset,
            'random': str(randomize).lower()
        }
        
        return self._make_request(f'/questions/category/{category}', params)
    
    def get_questions_by_difficulty(self,
                                   difficulty: str,
                                   limit: int = 100,
                                   offset: int = 0,
                                   randomize: bool = False) -> APIResponse:
        """
        Get questions from a specific difficulty level
        
        Args:
            difficulty: Difficulty level
            limit: Number of questions to return
            offset: Pagination offset
            randomize: Randomize results
        """
        params = {
            'limit': limit,
            'offset': offset,
            'random': str(randomize).lower()
        }
        
        return self._make_request(f'/questions/difficulty/{difficulty}', params)
    
    def get_categories(self) -> APIResponse:
        """Get all available categories"""
        return self._make_request('/categories')
    
    def get_difficulties(self) -> APIResponse:
        """Get all available difficulty levels"""
        return self._make_request('/difficulties')
    
    def get_levels(self) -> APIResponse:
        """Get all available game levels"""
        return self._make_request('/levels')
    
    def export_all_questions(self, include_answers: bool = True) -> APIResponse:
        """
        Export all questions as JSON
        
        Args:
            include_answers: Whether to include correct answers
        """
        params = {'include_answers': str(include_answers).lower()}
        return self._make_request('/export/json', params)
    
    def parse_questions(self, response: APIResponse) -> List[Question]:
        """
        Parse API response into Question objects
        
        Args:
            response: API response containing questions
            
        Returns:
            List of Question objects
        """
        if not response.success:
            return []
        
        questions = response.data.get('questions', [])
        return [
            Question(
                id=q.get('id', 0),
                question=q.get('question', ''),
                options=q.get('options', []),
                answer=q.get('answer', 0),
                difficulty=q.get('difficulty', ''),
                category=q.get('category', ''),
                level=q.get('level', 1),
                level_name=q.get('level_name', '')
            )
            for q in questions
        ]

# Example usage and testing
if __name__ == '__main__':
    # Example usage
    client = AWSTriviaAPIClient('http://localhost:5001')
    
    print("🚀 AWS Trivia API Client Example")
    print("=" * 50)
    
    # Health check
    print("\n1. Health Check:")
    health = client.health_check()
    if health.success:
        print(f"✅ API is healthy - {health.data.get('total_questions', 0)} questions available")
    else:
        print(f"❌ API health check failed: {health.error}")
    
    # Get API info
    print("\n2. API Information:")
    info = client.get_api_info()
    if info.success:
        data = info.data
        print(f"📊 Total Questions: {data.get('total_questions', 0)}")
        print(f"🏷️  Categories: {len(data.get('categories', []))}")
        print(f"📈 Difficulties: {len(data.get('difficulties', []))}")
        print(f"🎯 Levels: {len(data.get('levels', {}))}")
    
    # Get random questions
    print("\n3. Random Questions (3):")
    random_response = client.get_random_questions(count=3)
    if random_response.success:
        questions = client.parse_questions(random_response)
        for i, q in enumerate(questions, 1):
            print(f"\n   Question {i}: {q.question}")
            print(f"   Category: {q.category} | Difficulty: {q.difficulty} | Level: {q.level}")
            for j, option in enumerate(q.options):
                marker = "✓" if j == q.answer else " "
                print(f"   {marker} {chr(65+j)}. {option}")
    
    # Get questions by level
    print("\n4. Level 1 Questions (2):")
    level_response = client.get_questions_by_level(level=1, limit=2)
    if level_response.success:
        questions = client.parse_questions(level_response)
        for q in questions:
            print(f"   • {q.question} ({q.category})")
    
    # Get categories
    print("\n5. Available Categories:")
    categories_response = client.get_categories()
    if categories_response.success:
        categories = categories_response.data.get('categories', [])
        print(f"   {', '.join(categories)}")
    
    print("\n" + "=" * 50)
    print("✅ API Client Example Complete!")
